core/src/main/java/com/linecorp/armeria/client/limit/DefaultConcurrencyLimit.java

```diff
@@ -0,0 +1,181 @@
+ */
+
+package com.linecorp.armeria.client.limit;
+
+import static com.linecorp.armeria.client.limit.ConcurrencyLimitBuilder.noLimitFuture;
+
+import java.util.Queue;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.ConcurrentLinkedQueue;
+import java.util.concurrent.ScheduledFuture;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+import java.util.concurrent.atomic.AtomicReference;
+import java.util.function.Predicate;
+
+import com.google.common.annotations.VisibleForTesting;
+import com.google.common.base.MoreObjects;
+import com.spotify.futures.CompletableFutures;
+
+import com.linecorp.armeria.client.ClientRequestContext;
+import com.linecorp.armeria.client.UnprocessedRequestException;
+import com.linecorp.armeria.common.util.SafeCloseable;
+
+/**
+ * Concurrency settings that limits the concurrent number of active requests.
+ */
+final class DefaultConcurrencyLimit implements ConcurrencyLimit {
+
+    private final Predicate<? super ClientRequestContext> predicate;
+    private final int maxConcurrency;
+    private final int maxPendingAcquisitions;
+    private final long timeoutMillis;
+
+    private final Queue<PendingAcquisition> pendingAcquisitions = new ConcurrentLinkedQueue<>();
+    private final AtomicInteger acquiredPermits = new AtomicInteger();
+
+    DefaultConcurrencyLimit(Predicate<? super ClientRequestContext> predicate,
+                            int maxConcurrency, int maxPendingAcquisitions, long timeoutMillis) {
+        this.predicate = predicate;
+        this.maxConcurrency = maxConcurrency;
+        this.maxPendingAcquisitions = maxPendingAcquisitions;
+        this.timeoutMillis = timeoutMillis;
+    }
+
+    @VisibleForTesting
+    int acquiredPermits() {
+        return acquiredPermits.get();
+    }
+
+    @VisibleForTesting
+    int availablePermits() {
+        return maxConcurrency - acquiredPermits.get();
+    }
+
+    @Override
+    public CompletableFuture<SafeCloseable> acquire(ClientRequestContext ctx) {
+        if (!predicate.test(ctx)) {
+            return noLimitFuture;
+        }
+
+        if (availablePermits() == 0 && pendingAcquisitions.size() >= maxPendingAcquisitions) {
+            // Do not strictly calculate whether it exceeds the maxPendingAcquisitions using synchronized block.
+            return CompletableFutures.exceptionallyCompletedFuture(ExceedingMaxPendingException.get());
+        }
+
+        final CompletableFuture<SafeCloseable> future = new CompletableFuture<>();
+        final PendingAcquisition currentAcquire = new PendingAcquisition(future);
+        pendingAcquisitions.add(currentAcquire);
+        drain();
+
+        if (!currentAcquire.isRun() && timeoutMillis != 0) {
+            // Current acquire was not run. Schedule a timeout.
+            final ScheduledFuture<?> timeoutFuture = ctx.eventLoop().withoutContext().schedule(
+                    () -> future.completeExceptionally(
```

Comment:
Looks like the timed out task isn't being removed from the queue. Actually I guess removing would be too slow with either `ConcurrentLinkedQueue` or `ArrayBlockingQueue` both are O(n), Netty event loop has a custom linked list for scheduled futures that allows removing nodes in constant time (I think I might have added that a long time ago :-D).

Timeouts look like they should be rare, so I guess it should be OK to not remove until the next drain for memory reasons. But currently `pendingAcquisitions.size()` is used above for an important check so it not reflecting timed out acquisitions seems unsafe since timeout would only get checked on a drain - in above comment, I suggested adding `numPendingAcquisitions` and it's probably good to decrement here instead of waiting until drain.

Reply:
Though the problem with Netty is that there could be hundreds of thousands of timeouts in the event loop. If this number is usually relatively small than ok to call `remove`.

Reply:
How about maintaining a separate counter rather than using `size()`? then we don't really need to remove for each timeout but only when dequeueing.